bloom-works/guides-template
# Add CSS Linter

This changes is related to the [Setup Guides](https://github.com/bloom-works/guides-template/issues/5) task. It adds linting functionality for CSS files. The linter can be run at any point during development. It is also added to the workflow process (from PR click checks tab in top-middle of screen, click Pull Request on left side of screen, click pr_checks on left-side or middle of screen, click Run Linter in middle of screen). Prior to this change the run linter option in PR Check was a print statement as a placeholder. The image below shows a screenshot of the workflow with stylelint running and passing during the PR check of this current PR.

<img width="1022" alt="Screenshot 2023-07-04 at 11 15 27 AM" src="https://github.com/bloom-works/guides-template/assets/137226010/14a8f110-0cf2-4b0b-a479-c73053538fa7">

To make it more consumable for review, I will split up all the linter work into PRs associated to each linter and language. This review will include all changes for stylelint and CSS code.

### What was changed

- Added stylelint capabilities (packages, configuration)
- Updated sh file used in github workflows for pr check
- Add steps to run before PR to developer section of README
- Fixed changes in CSS files reported by linter (see notes below for full output of errors fixed)

### How to check change

1. Checkout branch and build project
2. Verify all changes reported by linter did not disrupt the project by running `npm run serve`
3. Verify there are no existing linter errors by using the path on PR above to see if pr checks passed for linter
4. Verify there are no existing linter errors by manually running the linter in the code. Navigate to the base of the repo and run `sh ./ci/linter.sh`. Below image shows an example of what the command line will look like when there are no errors.
<img width="302" alt="Screenshot 2023-07-04 at 10 57 50 AM" src="https://github.com/bloom-works/guides-template/assets/137226010/2968e1e4-7bee-4268-be56-538aa68a97d9">

5. Verify linter is working and/or check any style fixes I made in other files. For this you will use the same command as above and choose any of the files that have changed and reverse any changes made. I have give two examples below, but feel free to check any of the changes using the same process.

Example 1: In the below screenshot, I have changed the capitalization of the letter 'm' in 'Min' on line 70 of the layout.css file and then run the linter.  Note the linter now reports an error for _function-name_case_ with the message _Expected "Min" to be "min"_.
<img width="743" alt="Screenshot 2023-07-04 at 10 59 41 AM" src="https://github.com/bloom-works/guides-template/assets/137226010/3e8e4dad-21e0-4493-afcd-99082a00f772">

Example 2:  In the below screenshot, I have changed the numerical values from percentage to decimal (10% -> 0.1, 6% -> .06).  Note the linter now reports two errors for _alpha-value-notation_ with the message _Expected ".1" to be "10%"_ and _Expected ".06" to be "6%"_.
<img width="496" alt="Screenshot 2023-07-04 at 11 01 00 AM" src="https://github.com/bloom-works/guides-template/assets/137226010/2ee5735d-4603-467d-902e-9871c768fc92">


### Notes
I have included the stdout used to make all CSS changes for easier reviewing:

_includes/css/base/base.css
 27:21  ✖  Expected complex :not() pseudo-class notation  selector-not-notation

_includes/css/base/layout.css
  37:3   ✖  Expected empty line before declaration  declaration-empty-line-before
  61:3   ✖  Expected empty line before declaration  declaration-empty-line-before
  68:52  ✖  Expected "Min" to be "min"              function-name-case
 147:3   ✖  Expected empty line before comment      comment-empty-line-before

_includes/css/base/normalize.css
 366:1  ✖  Expected empty line before comment  comment-empty-line-before

_includes/css/base/root.css
  8:51  ✖  Expected whitespace before "*/"               comment-whitespace-inside
 39:3   ✖  Unexpected empty line before custom property  custom-property-empty-line-before
 42:3   ✖  Unexpected empty line before custom property  custom-property-empty-line-before

_includes/css/base/typography.css
  55:1   ✖  Unexpected duplicate selector "h1, h2, h3, h4, h5, h6, .h1, .h2, .h3, .h4, .h5, .h6", first used at line 50  no-duplicate-selectors
  91:3   ✖  Expected empty line before custom property                                                                   custom-property-empty-line-before
  97:3   ✖  Expected empty line before custom property                                                                   custom-property-empty-line-before
 161:26  ✖  Expected modern color-function notation                                                                      color-function-notation
 177:1   ✖  Expected selector "figcaption" to come before selector "blockquote + figcaption"                             no-descending-specificity
 184:3   ✖  Expected empty line before declaration                                                                       declaration-empty-line-before

_includes/css/components/button.css
  9:30  ✖  Expected modern color-function notation     color-function-notation
  9:44  ✖  Expected "0.1" to be "10%"                  alpha-value-notation
 10:30  ✖  Expected modern color-function notation     color-function-notation
 10:44  ✖  Expected "0.06" to be "6%"                  alpha-value-notation
 48:3   ✖  Expected empty line before custom property  custom-property-empty-line-before

_includes/css/components/footer.css
 28:1  ✖  Unexpected duplicate selector ".footer", first used at line 3                  no-duplicate-selectors
 40:1  ✖  Unexpected duplicate selector ".footer::before", first used at line 33         no-duplicate-selectors
 96:1  ✖  Expected selector ".footer-icons a" to come before selector ".footer a:focus"  no-descending-specificity
 96:1  ✖  Expected selector ".footer-icons a" to come before selector ".footer a:hover"  no-descending-specificity

_includes/css/components/pagination.css
 18:3  ✖  Expected empty line before custom property  custom-property-empty-line-before

_includes/css/components/resources.css
  7:3   ✖  Expected empty line before declaration   declaration-empty-line-before
 30:3   ✖  Expected empty line before declaration   declaration-empty-line-before
 49:3   ✖  Unexpected duplicate "display"           declaration-block-no-duplicate-properties
 75:25  ✖  Expected modern color-function notation  color-function-notation
 75:39  ✖  Expected "0.06" to be "6%"               alpha-value-notation
 90:27  ✖  Expected modern color-function notation  color-function-notation
 90:41  ✖  Expected "0.1" to be "10%"               alpha-value-notation

_includes/css/components/tables.css
 29:19  ✖  Expected "context" media feature range notation  media-feature-range-notation
 83:19  ✖  Expected "context" media feature range notation  media-feature-range-notation

34 problems (34 errors, 0 warnings)
